feat(cli): add `tsconfig` file detection based on the framework

sadeghbarati opened the pull request:
<!---☝️ PR title should follow conventional commits (https://conventionalcommits.org) -->

### 🔗 Linked issue


### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [x] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

- ~~Add `tsConfigPath` options to the `components.json` options schema~~
- Check for frameworks for detecting the tsconfig file
- Add main shadcn-ui CLI changes
- Use `c12` again for loading component.json/json5/js/ts  (it is working in Nuxt projects with empty or full components directory)
- Enable `shims` and add nodeProtocol: strip to remove `node:` protocol from the build


### 📸 Screenshots (if appropriate)

<!-- Add screenshots to help explain the change. -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

sadeghbarati committed:
feat(cli): add back `tsconfigPath` option to component.json

add main shadcn-ui cli changes
use c12 again for loading component.json/json5/js/ts

sadeghbarati committed:
fix(cli): use `shim` and remove node protocol from the dist

export mcp in package.json

sadeghbarati committed:
refactor(cli): `tsconfigPath` -> `tsConfigPath`

sadeghbarati committed:
chore: replace remaining` node:path` with `pathe` module

sadeghbarati committed:
feat: implement registry safe path

from main shadcn

sadeghbarati committed:
feat: fix safe target and

from shadcn

sadeghbarati committed:
feat(cli): add support for universal registry item

from shadcn

sadeghbarati committed:
fix: handling of themes in registry

from shadcn

sadeghbarati committed:
fix: universal item files type

from shadcn

zernonia commented:
Yo @sadeghbarati ! Thanks for adding this field back into the config. I wonder why this wasn't a request on actual `ui.shadcn`? Perhaps it would be better if we have the same [`monorepo`](https://ui.shadcn.com/docs/monorepo#importing-components) support?

sadeghbarati commented:
@zernonia 

I think the monorepo CLI code is already here

We might need to remove the tsConfigPath option 

But in Nuxt 4, we have an issue that this option may resolve it  #1340

tsConfigPath is optional 

---

Monorepo is too complicated people is just need to use normal directory for component registry

sadeghbarati committed:
Merge remote-tracking branch 'origin/dev' into configure-tsconfigPath-in-component-dot-json

sadeghbarati committed:
feat: update new shadcn CLI changes

from shadcn

sadeghbarati committed:
Merge branch 'dev' into configure-tsconfigPath-in-component-dot-json

sadeghbarati committed:
feat: update

sadeghbarati committed:
fix: resolvedPaths: required and ofetch response

sadeghbarati commented:
We have multiple choices

1. make main tsconfig.json/tsconfig.app.json automatically detectable based on framework and framework version (not implemented ❌)
2. add new flag on init and add CLI like --tsconfig-path  (not implemented ❌)
3. `tsConfigPath` in components.json (implemented ✅)

What we should do? which way is more convenient

sadeghbarati committed:
refactor: get correct tsconfig file based on the nuxt versions

sadeghbarati committed:
refactor: remove tsConfigPath option and add auto detect tsconfig files based on the framework

sadeghbarati committed:
Merge branch 'dev' into configure-tsconfigPath-in-component-dot-json

sadeghbarati committed:
chore: update lockfile

sadeghbarati committed:
refactor: add inertia detection

sadeghbarati commented:
Hi, @zernonia, could you please check this PR? Thank you ❤️ 

I removed the `tsConfigPath` option and added the tsconfig detection from the existing projectInfos utilities

The PR also has the latest changes from the main shadcn ui too


There is a lot of room for improvement for detection, but let's start small

ZTL-UwU commented on packages/cli/src/utils/get-project-info.ts:
```diff
@@ -41,22 +38,68 @@ const TS_CONFIG_SCHEMA = z.object({
+  })
+
+  // Check for Nuxt
+  if (configFiles.find(file => file.startsWith('nuxt.config.'))) {
+    const isUsingAppDir = await fs.pathExists(path.resolve(cwd, 'app'))
+    return isUsingAppDir ? FRAMEWORKS.nuxt4 : FRAMEWORKS.nuxt3
+  }
```
@sadeghbarati There is a problem with this detection approach. Having the `app` folder does not necessarily mean it is nuxt4, since nuxt3 (with `compatibilityVersion: 4`) also uses the `app` folder structure, and nuxt4 has downward compatibility for the old file structure as well.

Instead, we should probably check the `package.json` for the exact nuxt version using semver, which is what nuxi is doing.

https://github.com/nuxt/cli/blob/77b68177d574e302c91dfc86137e9b9aaa95343f/packages/nuxi/src/utils/versions.ts#L1-L12

```ts
import { readPackageJSON } from 'pkg-types'
import { coerce } from 'semver'

export async function getNuxtVersion(cwd: string, cache = true) {
  const pkg = await readPackageJSON(cwd)
  const pkgDep = pkg?.dependencies?.nuxt || pkg?.devDependencies?.nuxt
  return (pkgDep && coerce(pkgDep)?.version) || '3.0.0'
}
```

sadeghbarati replied:
Thanks ❤️

sadeghbarati committed:
Merge remote-tracking branch 'origin/dev' into configure-tsconfigPath-in-component-dot-json

sadeghbarati committed:
refactor: improve `nuxt` version detection

sadeghbarati committed:
fix: update nuxt labels for clarity

sadeghbarati committed:
fix(cli): fix --defaults options

from shadcn

sadeghbarati committed:
feat(cli): mcp init

from shadcn

sadeghbarati committed:
chore: replace node:path with pathe

sadeghbarati committed:
chore: update tests

sadeghbarati committed:
chore: update tests

sadeghbarati committed:
chore: bump mcp sdk

sadeghbarati committed:
feat: allow empty files items

from shadcn

sadeghbarati committed:
feat: update handling of import and apply at rules

from shadcn

sadeghbarati commented:
Tests need to be updated. I'm not a pro at it 😭 
Or just merge it without test files, np all good 😁

sadeghbarati committed:
Merge branch 'dev' into configure-tsconfigPath-in-component-dot-json

sadeghbarati committed:
feat(cli): add support for registries index
feat(cli): implement recursive registry namespaces

from shadcn

sadeghbarati committed:
chore: update dependencies and improve framework detection

mayar-deeb-xi commented:
I came from one of the linked issue..... , any plans to merge this :)

sadeghbarati commented:
@mayar-deeb-xi  Hi, For now use it like 


1. Based on your package manager for example `pnpm`

```
pnpm add https://pkg.pr.new/shadcn-vue@1330 -D
```

2. Add new script to scripts in `package.json` like


```json
{
    ...
	"scripts": {
		"dev": "vite",
		"shadcn-vue": "shadcn-vue"
	},
	...
}
```

3. run it in your project directory without `pnpm dlx` or `npx`

```
pnpm run shadcn-vue init/add
```

sadeghbarati committed:
chore: update deps

sadeghbarati committed:
chore: update

sadeghbarati committed:
Merge branch 'dev' into configure-tsconfigPath-in-component-dot-json

sadeghbarati commented:
Self-merging for merging these changes to the [`v4-docs`](https://github.com/unovue/shadcn-vue/tree/v4-docs) branch
